src/read/mod.rs

```diff
@@ -38,7 +41,11 @@ where
     }
 
     reader.seek(SeekFrom::Start(eocdr.cent_dir_offset.into())).await?;
-    let entries = crate::read::cd(&mut reader, eocdr.num_of_entries.into()).await?;
+
+    // To avoid lots of small reads to `reader` when parsing the central directory, we use a BufReader that can read the whole central directory at once.
+    // Because `eocdr.size_cent_dir` is a u32, we use MAX_CD_BUFFER_SIZE to prevent very large buffer sizes.
+    let buf = BufReader::with_capacity(std::cmp::min(eocdr.size_cent_dir as _, MAX_CD_BUFFER_SIZE), reader);
```

Comment:
I don't think this is correct.

Suppose that after all data from `ZipFile` is consumed, the `BufReader` still has some data in its own buffer that belongs to the next `ZipFile`, then the data will be discarded.

The only way to do this properly is use a global `BufReader` for the entire `ZipFileReader`.

Reply:
I think the current implementation is okay because `reader` is `AsyncSeek` so even if there is data in the `BufReader` after parsing the central directory, it isn't consumed or discarded in a way that makes it inaccessible in the future.

Also, as far as I can tell, nothing depends on the position of `reader` being in a specific location after `file` returns.

When something does depend on a specific positon, it seems to calls seek first. For example:

https://github.com/Majored/rs-async-zip/blob/07721c34942c129156df4e7810f4b520b719e6d9/src/read/seek.rs#L65-L76

Also what do you mean by "all data from `ZipFile` is consumed"? A `ZipFile` doesn't exist yet at this point in the code, right? The BufReader is dropped by the time a `ZipFile` is returned by `fiile()`

Reply:
> Also what do you mean by "all data from ZipFile is consumed"? A ZipFile doesn't exist yet at this point in the code, right? The BufReader is dropped by the time a ZipFile is returned by fiile()

Sorry I thought the `BufReader` is returned in `ZipFile`, turns out I was wrong, but my point about `BufReader` is still valid.

> I think the current implementation is okay because reader is AsyncSeek so even if there is data in the BufReader after parsing the central directory, it isn't consumed or discarded in a way that makes it inaccessible in the future.
> 
> Also, as far as I can tell, nothing depends on the position of reader being in a specific location after file returns.

If that is `AsyncSeek`, then it's certainly ok, but there's also the stream implementation that's currently removed, not sure how this will interact with that.

Reply:
Yeah the `file` function has an `AsyncSeek` bound on `reader`. It needs it in order to seek to the central directory (line 43).

RE stream implementation: The central directory is at the end of a zip file so it's usually not used for (non-seekable) streaming decompression. (Streaming decompression implementations use the local file headers instead of the central directory.) Because of that, if the stream implementation comes back, it likely won't use this code anyway

Reply:
Thanks for the explanation!